casper/src/main/scala/coop/rchain/casper/engine/BlockReceiver.scala

```diff
@@ -0,0 +1,87 @@
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import coop.rchain.blockstorage.blockStore.BlockStore
+import coop.rchain.blockstorage.casperbuffer.CasperBufferStorage
+import coop.rchain.blockstorage.dag.BlockDagRepresentation
+import coop.rchain.blockstorage.util.DoublyLinkedDag
+import coop.rchain.comm.PeerNode
+import coop.rchain.models.BlockHash.BlockHash
+
+sealed trait RecvStatus
+case object Receiving   extends RecvStatus // Block receiving started
+case object Received    extends RecvStatus // Block fully received
+case object Unvalidated extends RecvStatus // Block received but not validated (put in BlockStore)
+case object Validated   extends RecvStatus // Block received and validated (ready to put in BlockMetadataStore/DAG)
+case object Invalid     extends RecvStatus // Block received but invalid
+
+final case class BlockReceiverState[MId](state: Map[MId, RecvStatus])
+
+final case class BlockReceiver[F[_]: Sync](
+    ref: Ref[F, BlockReceiverState[BlockHash]],
+    blockRetriever: BlockRetriever[F],
+    casperBuffer: CasperBufferStorage[F]
+) {
+  def contains(m: BlockHash): F[Boolean] = ref.get.map(_.state.contains(m))
+
+  def addIfNotExists(m: BlockHash): F[Unit] =
+    ref.update { st =>
+      if (st.state.contains(m)) {
+        st
+      } else {
+        BlockReceiverState(st.state + (m -> Receiving))
+      }
+    }
+
+  def set(m: BlockHash, status: RecvStatus): F[Unit] = ref.update { st =>
+    BlockReceiverState(st.state + (m -> status))
+  }
+  def filter(status: RecvStatus): F[Set[BlockHash]] =
+    ref.get.map(_.state.filter { case (_, st) => st == status }.keySet)
+
+  // BlockRetriever
+  def admitHash(
+      hash: BlockHash,
+      peer: Option[PeerNode] = None,
+      admitHashReason: BlockRetriever.AdmitHashReason
+  ): F[BlockRetriever.AdmitHashResult] = blockRetriever.admitHash(hash, peer, admitHashReason)
+
+  // CasperBuffer
+  def getPendants: F[Set[BlockHash]] = casperBuffer.getPendants
+  def bufferRemove(blockHash: BlockHash): F[Unit] = {
+    val _ = casperBuffer.remove(blockHash)
+    // Hash removes from CasperBuffer only if it validated
+    set(blockHash, Validated)
+  }
+  def bufferContains(blockHash: BlockHash): F[Boolean] = casperBuffer.contains(blockHash)
+  def toDoublyLinkedDag: F[DoublyLinkedDag[BlockHash]] = casperBuffer.toDoublyLinkedDag
+}
+
+object BlockReceiver {
+  def create[F[_]: Sync](
+      blockRetriever: BlockRetriever[F],
+      casperBuffer: CasperBufferStorage[F]
+  ): F[BlockReceiver[F]] =
+    for {
+      ref <- Ref[F].of(BlockReceiverState(Map.empty[BlockHash, RecvStatus]))
+    } yield BlockReceiver(ref, blockRetriever, casperBuffer)
+
+  def create[F[_]: Sync /*: Log*/ ](
+      blockRetriever: BlockRetriever[F],
+      casperBuffer: CasperBufferStorage[F],
+      blockStore: BlockStore[F],
+      blockDagRepresentation: BlockDagRepresentation[F]
+  ): F[BlockReceiver[F]] =
+    for {
+      blockHashes <- blockStore.toMap >>= { _.keySet.pure }
```

Comment:
`toMap` on the block store is very heavy operation, it will load all blocks which can be tens of GBs.

We don't need to load all blocks at the beginning, BlockReceiver just needs to resolve what to do with newly recived block hash and block. Is it ready for the next stage (ready to validation/DAG), missing dependencies or it should be ignored.

On the beginning casper buffer must be processed because it contains blocks that are downloaded but not yet in the DAG.

Reply:
@tgrospic, I decided to use hashes from `CasperBuffer` as `Unvalidated` hashes. `CasperBufferStorage` hasn't got functions to collect all stored hashes so I added `nodes` function for that. Or maybe it should be named `blocks` or `messages`?